js/background.js

```diff
@@ -107,85 +103,109 @@ chrome.contextMenus.create({
-          }
-
-          return upgradeStatus;
-      }
-      */
-
+    function (requestData) { 
+
+        let tabId = requestData.tabId;
+
+        // Add ATB for DDG URLs
+        let ddgAtbRewrite = ATB.redirectURL(requestData);
+        if (ddgAtbRewrite) return ddgAtbRewrite;
+
+        // Skip requests to background tabs
+        if (tabId === -1) { return }
+
+        let thisTab = tabManager.get(requestData);
+
+        // For main_frame requests: create a new tab instance whenever we either
+        // don't have a tab instance for this tabId or this is a new requestId.
+        if (requestData.type === "main_frame") {
+            if (!thisTab || (thisTab.requestId !== requestData.requestId)) {
+              thisTab = tabManager.create(requestData);
+            }
+        }
+        else {
+
+            /**
+             * Check that we have a valid tab
+             * there is a chance this tab was closed before
+             * we got the webrequest event
+             */
+            if (!(thisTab && thisTab.url && thisTab.id)) return
+
+            /**
+             * Tracker blocking 
+             * If request is a tracker, cancel the request 
+             */
+
+            chrome.runtime.sendMessage({"updateTrackerCount": true});
+
+            var tracker =  trackers.isTracker(requestData.url, thisTab.url, thisTab.id, requestData);
+
+            if (tracker) {
+                // record all tracker urls on a site even if we don't block them
+                thisTab.site.addTracker(tracker);
+
+                // record potential blocked trackers for this tab
+                thisTab.addToTrackers(tracker);
+
+                // Block the request if the site is not whitelisted
+                if (!thisTab.site.whitelisted) {
+                    thisTab.addOrUpdateTrackersBlocked(tracker);
+                    chrome.runtime.sendMessage({"updateTrackerCount": true});
+
+                    // update badge icon for any requests that come in after
+                    // the tab has finished loading
+                    if (thisTab.status === "complete") thisTab.updateBadgeIcon()
+
+                    console.info( utils.extractHostFromURL(thisTab.url)
+                                 + " [" + tracker.parentCompany + "] " + tracker.url);
+
+                    if (tracker.parentCompany !== 'unknown') Companies.add(tracker.parentCompany)
+
+                    // tell Chrome to cancel this webrequest
+                    return {cancel: true};
+                }
+            }
+        }
+        
+        /**
+         * HTTPS Everywhere rules
+         * If an upgrade rule is found, request is upgraded from http to https 
+         */
+
+         if (!thisTab.site) return
+
+        // Avoid redirect loops
+        if (thisTab.httpsRedirects[requestData.requestId] >= 7) {
+            console.log('HTTPS: cancel request. redirect limit exceeded for url: \n' + requestData.url)
+            return {cancel: true}
```

Comment:
Do we want to downgrade instead of cancel in this case?

Reply:
Hm, yeah, good catch.